Fix CPH_Merger.py

neverrend opened the pull request:
The merger was stripping the first char of the word "class" when it was ran. Added a one to the len so it properly merged the files.

PS. Thanks for fixing that escaping on the + chars it was driving me bonkers during tests.

Commit:
Fix CPH_Merger.py	The merger was stripping the first char of

the word "class" when it was ran. Added a one to the len so it properly
merged the files.

elespike commented:
Hi, @neverrend!

Thanks for the PR. Something's wonky here, though.

I can't reproduce the issue, and if `CPH_Merger.py` were broken, that would mean that loading the extension in Burp Suite directly from their [BApp store](https://portswigger.net/bappstore/a0c0cd68ab7c4928b3bf0a9ad48ec8c7) would fail.

Would you mind breaking down how you came across this behavior?

neverrend commented:
Yeah it’s definitely strange. I pulled it down and ran the merger. When I put the merged .py file into burp I got an error with a stack trace. When I went to the offending line it had “lass” rather “class” at the start of where each file was merged in. I manually went in and fixed the errors and it loaded just fine. I did some diagnosing and it seemed that len(line) was reading out to 16. The len(line)+1 set it all back to normal and next time I ran it I got a merged file that loaded into burp just fine.

elespike commented:
Which version of Python did you use to run `CPH_Merger.py`?
I'm using `2.7`, since that's what Jython supports.

Can't use Python 3 until Jython supports it. :neutral_face: 

Here's a recording I just made, everything works ok:
https://asciinema.org/a/w21VMT52GemFvsYiu7jIsgbLL

neverrend commented:
I’m pretty sure I was 2.7 but I’m not near the rig that this occurred on. I’ll get the numbers and reproduce it and get back to you in the morning. It’s curious for sure.

Commit:
Fix CPH_Merger.py 	Revised fix for this script.

Some further bug testing has shown that this is an OS specific issue
so a ternary operator was added.

neverrend commented:
Hello @elespike,

I have figured out the issue after some testing of my own. It seems to be an OS issue rather than a code issue. I am running the same version of python ```2.7.16``` and still ran into the same issue where "class" was being truncated. I did the same on my Linux box with the same ```2.7.16``` version and it ran like a charm, no truncated lines. I am going to include a revised fix to check for OS before it sets those line length values. 

Below is a video showing the behavior before and after the fix:
https://streamable.com/dvam1

See the updated fix here: [```735e276```](https://github.com/elespike/burp-cph/pull/42/commits/735e2768696a507ebcfc28f94e908a58cd057c5b)

Cheers!

elespike commented:
Wow, good catch! I didn't even think that the OS could be the culprit. CPH in the beginning was written and tested on Windows.

Let me run some tests to see if I can understand the root cause of this, and I'll add to the PR if needed. =)

Thanks a lot for your help!

neverrend commented:
I believe it has something to do with the way the SEEK_CUR is used but I haven’t had the time to dig into the actual why. Just enough to know it was an OS issue. 

I only caught it cause for some reason your latest commit wasn’t pushed onto my version of CPH and the regex was appending two base64 encoded values rather than replacing one with the former. Once I figured it was a bug I went looking for the repo and sure enough you had just issued a commit for the issue correcting the escaping + values. I pulled it down and ran the merger and got the stack trace from burp which sent me on the hunt. 

Anyway if you figure it out I’d love to hear what was actually happening with it. Or if I get some free time I’ll also dig into it to see what’s going on and update back on here.

elespike commented:
I'm still not 100% sure of what's going on, but I did identify an appropriate fix. Check out the https://github.com/elespike/burp-cph/tree/xplatf-merger branch for details.

In summary, it seems that file seeking operations can behave weirdly if the line endings don't match what the OS expects. To make matters worse, `git` can be configured to checkout files with their original line endings, or to automatically match line endings to those expected by the current OS - this means that checking the platform wouldn't be enough.

Fortunately, these issues can be avoided by simply reading and writing in binary mode. For example:
`open('my_file', 'rb')` - note `rb` for "read, binary"
This was the main change which fixes this bug.

I'll wait for you to verify it on your end before merging anything.

elespike commented:
@neverrend, ping me when you've had a look at https://github.com/elespike/burp-cph/tree/xplatf-merger. I'll probably create a PR from that one once you've verified it.
